crypto/evp/evp_exchange.c

```diff
@@ -0,0 +1,350 @@
+                break;
+            exchange->dupctx = OSSL_get_OP_exchange_dupctx(fns);
+            break;
+        case OSSL_FUNC_EXCHANGE_SET_PARAMS:
+            if (exchange->set_params != NULL)
+                break;
+            exchange->set_params = OSSL_get_OP_exchange_set_params(fns);
+            break;
+        }
+    }
+    if (fncnt != 4) {
+        /*
+         * In order to be a consistent set of functions we must have at least
+         * a complete set of "exchange" functions: init, derive, newctx,
+         * and freectx. The dupctx, set_peer and set_params functions are
+         * optional.
+         */
+        EVP_EXCHANGE_free(exchange);
+        EVPerr(EVP_F_EVP_EXCHANGE_FROM_DISPATCH, EVP_R_INVALID_PROVIDER_FUNCTIONS);
+        return NULL;
+    }
+
+    return exchange;
+}
+
+void EVP_EXCHANGE_free(EVP_EXCHANGE *exchange)
+{
+    if (exchange != NULL) {
+        int i;
+
+        CRYPTO_DOWN_REF(&exchange->refcnt, &i, exchange->lock);
+        if (i > 0)
+            return;
+        ossl_provider_free(exchange->prov);
+        CRYPTO_THREAD_lock_free(exchange->lock);
+        OPENSSL_free(exchange);
+    }
+}
+
+int EVP_EXCHANGE_up_ref(EVP_EXCHANGE *exchange)
+{
+    int ref = 0;
+
+    CRYPTO_UP_REF(&exchange->refcnt, &ref, exchange->lock);
+    return 1;
+}
+
+EVP_EXCHANGE *EVP_EXCHANGE_fetch(OPENSSL_CTX *ctx, const char *algorithm,
+                                 const char *properties)
+{
+    return evp_generic_fetch(ctx, OSSL_OP_EXCHANGE, algorithm, properties,
+                             evp_exchange_from_dispatch,
+                             (int (*)(void *))EVP_EXCHANGE_up_ref,
+                             (void (*)(void *))EVP_EXCHANGE_free);
+}
+
+int EVP_PKEY_derive_init_ex(EVP_PKEY_CTX *ctx, EVP_EXCHANGE *exchange)
+{
+    int ret;
+    OSSL_PARAM *param = NULL;
+    size_t paramsz = 0;
+
+    ctx->operation = EVP_PKEY_OP_DERIVE;
+
+    if (ctx->engine != NULL)
+        goto legacy;
+
+    if (exchange != NULL) {
+        if (!EVP_EXCHANGE_up_ref(exchange))
+            goto err;
+    } else {
+        int nid = ctx->pkey != NULL ? ctx->pkey->type : ctx->pmeth->pkey_id;
+
+        /*
+         * TODO(3.0): Check for legacy handling. Remove this once all all
+         * algorithms are moved to providers.
+         */
+        if (ctx->pkey != NULL) {
+            switch (ctx->pkey->type) {
+            case EVP_PKEY_DH:
+                break;
+            default:
+                goto legacy;
+            }
+            /*
+             * If the EVP_PKEY has not been set, then this is a legacy use of
+             * the key exchange functions (scrypt/tls1_prf/hkdf). They're really
+             * KDFs. We don't want explicit fetches to find these "key exchange"
+             * algorithms, so we add the "internal" property.
+             */
+            exchange = EVP_EXCHANGE_fetch(NULL, OBJ_nid2sn(nid),
+                                          ctx->pkey == NULL ? "internal" : NULL);
```

Comment:
I have absolutely no idea what this "internal" property really does..
I cant see this property anywhere.

Reply:
There are a number of algorithms (such as HKDF) which (for historical reasons) have been forced into the "derive" API. They have nothing to do with key exchange - they are KDFs, and are now available via the EVP_KDF APIs. For backwards compatibility we need to be able to support them when doing implicit fetches. But I don't think we should encourage their use in an explicit fetch. For that reason I thought we might add an additional "internal" property only for use when doing implicit fetching. I'd expect those "key exchange" algorithms which are really KDFs to have the internal property, but not have the "default" property.

It's not currently used because I've only brought DH across as part of this PR. I expect future PRs to pull across the other algorithms.

If this is controversial I could drop this from this PR, and reintroduce it when we bring those algorithms across.

Reply:
Even a TODO above it would be fine..

Reply:
TODO added

Reply:
I don't think we should have a behaviour difference like this between implicit and explicit handling - that is asking for issues. Basically the decision on implicit/explicit shouldn't remove something that is available on implicit simply because the caller is using the explicit interface. That inconsistency is (in my view) substantially more concerning than someone using a less desirable interface.

Reply:
Shouldn't the code just fail in this situation?  This is the _key exchange_ code and it's considering (the unrelated) KDFs.  That seems wrong.

Would a better way be having the PKEY legacy support make the decision to go to either the key exchange code or the KDF code?

Reply:
This is the legacy support code....it shouldn't fail, it should dispatch to the "correct" API.

Reply:
Since this is controversial, I've just removed this for now. We can think more about what to do with this when we move the KDFs that use the "derive" API into the providers in a later PR.